File: packages/core/src/engine/fees.ts
import type { FeePolicy } from '../config/policy';
import { assertU64 } from '../math/fixed';
import { ErrorCode, fail } from './errors';

/**
 * Fee engine (ECONOMICS §3). Micro-USDC integers throughout.
 */

/** fee_required = notional × fee_bps / 10_000 (floor). */
export function feeRequired(notionalUsdc: bigint, feeBps: number): bigint {
  if (notionalUsdc < 0n) fail(ErrorCode.InvalidParams, 'negative notional');
  return (notionalUsdc * BigInt(feeBps)) / 10_000n;
}

/** Client-side charge for USDC entries: never below 0.50 % of USDC actually spent. */
export function feeCharged(notionalUsdc: bigint, usdcIn: bigint, feeBps: number): bigint {
  const a = feeRequired(notionalUsdc, feeBps);
  const b = feeRequired(usdcIn, feeBps);
  return a > b ? a : b;
}

export interface FeeSplitResult {
  toPool: bigint;
  toProtocol: bigint;
  toCreator: bigint;
  /** Where the creator share actually went. */
  creatorRoutedTo: 'Creator' | 'Protocol' | 'RewardPool';
}

/**
 * Split a paid fee. The protocol share absorbs rounding so
 * `toPool + toProtocol + toCreator == feePaid` exactly.
 */
export function splitFee(feePaid: bigint, policy: FeePolicy): FeeSplitResult {
  assertU64(feePaid, 'fee_paid');
  const pool = (feePaid * BigInt(policy.split.rewardPoolBps)) / 10_000n;
  const creator = (feePaid * BigInt(policy.split.creatorBps)) / 10_000n;
  const protocol = feePaid - pool - creator;
  switch (policy.creatorShareTarget) {
    case 'Creator':
      return { toPool: pool, toProtocol: protocol, toCreator: creator, creatorRoutedTo: 'Creator' };
    case 'Protocol':
      return {
        toPool: pool,
        toProtocol: protocol + creator,
        toCreator: 0n,
        creatorRoutedTo: 'Protocol',
      };
    case 'RewardPool':
      return {
        toPool: pool + creator,
        toProtocol: protocol,
        toCreator: 0n,
        creatorRoutedTo: 'RewardPool',
      };
  }
}

File: programs/tribe_arena/src/engine/fees.rs
//! Fee engine — transcription of `packages/core/src/engine/fees.ts` (ECONOMICS §3).

use crate::engine::math::to_u64;
use crate::state::{creator_target, FeePolicy};
use anchor_lang::prelude::*;

/// `fee_required = notional × fee_bps / 10_000` (floor).
pub fn fee_required(notional_usdc: u64, fee_bps: u16) -> Result<u64> {
    to_u64((notional_usdc as u128) * (fee_bps as u128) / 10_000)
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct FeeSplit {
    pub to_pool: u64,
    pub to_protocol: u64,
    pub to_creator: u64,
}

/// Split a paid fee; the protocol share absorbs rounding so the three parts
/// sum to `fee_paid` exactly. `target` routes the creator share.
pub fn split_fee(fee_paid: u64, policy: &FeePolicy, target: u8) -> Result<FeeSplit> {
    let f = fee_paid as u128;
    let pool = to_u64(f * policy.reward_pool_bps as u128 / 10_000)?;
    let creator = to_u64(f * policy.creator_bps as u128 / 10_000)?;
    let protocol = fee_paid - pool - creator;
    Ok(match target {
        creator_target::PROTOCOL => FeeSplit {
            to_pool: pool,
            to_protocol: protocol + creator,
            to_creator: 0,
        },
        creator_target::REWARD_POOL => FeeSplit {
            to_pool: pool + creator,
            to_protocol: protocol,
            to_creator: 0,
        },
        _ => FeeSplit {
            to_pool: pool,
            to_protocol: protocol,
            to_creator: creator,
        },
    })
}
